File: src/test/java/lambdaspecial/java/enroll/EnrollmentInfoTest.java
package lambdaspecial.java.enroll;

import lambdaspecial.java.enroll.domain.ClassInfo;
import lambdaspecial.java.enroll.domain.Student;
import lambdaspecial.java.enroll.domain.StudentClass;
import lambdaspecial.java.enroll.dto.EnrollmentInfo;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

public class EnrollmentInfoTest {

    List<Student> students = Arrays.asList(
            new Student(20231001, "홍길동", "강남구"),
            new Student(20231002, "이순신", "종로구"),
            new Student(20231003, "강감찬", "서초구"),
            new Student(20231004, "유관순", "마포구")
    );

    List<ClassInfo> classes = Arrays.asList(
            new ClassInfo(20231001, "수학", "김철수"),
            new ClassInfo(20231002, "과학", "이영희"),
            new ClassInfo(20231003, "영어", "박영수")
    );

    List<StudentClass> studentClasses = Arrays.asList(
            new StudentClass(20231001, 20231001),
            new StudentClass(20231002, 20231002),
            new StudentClass(20231003, 20231001),
            new StudentClass(20231004, 20231003)
    );

    @Test
    public void testJoin() {
        // 목표: SQL 쿼리와 동일한 결과 얻기
        // SELECT S.S_NO, S.NAME, S.ADDRESS, C.CLASS_NAME, C.TEACHER
        // FROM STUDENT S, STUDENT_CLASS SC, CLASS C
        // WHERE S.S_NO = SC.S_NO AND SC.C_NO = C.C_NO AND S.NAME = '홍길동';

        students.stream()
                // 1. S.NAME = '홍길동' 조건으로 먼저 필터링 (성능상 이점)
                .filter(student -> student.getName().equals("홍길동"))
                // 2. STUDENT 와 STUDENT_CLASS 조인 (S.S_NO = SC.S_NO)
                // flatMap을 사용하여 각 학생에 대해 매칭되는 수강 정보를 스트림으로 만듦
                .flatMap(student -> studentClasses.stream()
                        .filter(sc -> sc.getsNo() == student.getsNo())
                        // 조인된 결과를 임시 객체나 새로운 스트림 요소로 만듦
                        // 여기서는 Student와 StudentClass의 cNo를 함께 가지는 Pair 같은 것을 만들거나,
                        // 바로 다음 조인을 위해 cNo를 가지고 ClassInfo와 매칭할 수 있도록 함
                        .map(sc -> new Object[]{student, sc.getcNo()}) // 임시 배열로 학생과 수업번호 전달
                )
                // 3. 이전 조인 결과와 CLASS 조인 (SC.C_NO = C.C_NO)
                .flatMap(studentAndClassNo -> { // studentAndClassNo[0]는 Student, studentAndClassNo[1]는 cNo
                    Student student = (Student) studentAndClassNo[0];
                    int cNo = (Integer) studentAndClassNo[1];
                    return classes.stream()
                            .filter(cls -> cls.getcNo() == cNo)
                            .map(cls -> new EnrollmentInfo(
                                    student.getsNo(),
                                    student.getName(),
                                    student.getAddress(),
                                    cls.getClassName(),
                                    cls.getTeacher()
                            ));
                })
                .forEach(System.out::println);
    }
}

File: src/test/kotlin/lambdaspecial/kotlin/enroll/EnrollmentInfoTest.kt
package lambdaspecial.kotlin.enroll

import lambdaspecial.kotlin.enroll.domain.ClassInfo
import lambdaspecial.kotlin.enroll.domain.Student
import lambdaspecial.kotlin.enroll.domain.StudentClass
import lambdaspecial.kotlin.enroll.dto.EnrollmentInfo
import kotlin.test.Test

public class EnrollmentInfoTest {

    val students = listOf(
        Student(20231001, "홍길동", "강남구"),
        Student(20231002, "이순신", "종로구"),
        Student(20231003, "강감찬", "서초구"),
        Student(20231004, "유관순", "마포구")
    )

    val classes = listOf(
        ClassInfo(20231001, "수학", "김철수"),
        ClassInfo(20231002, "과학", "이영희"),
        ClassInfo(20231003, "영어", "박영수")
    )

    val studentClasses = listOf(
        StudentClass(20231001, 20231001),
        StudentClass(20231002, 20231002),
        StudentClass(20231003, 20231001),
        StudentClass(20231004, 20231003)
    )

    @Test
    fun testJoin() {
        // 목표 SQL:
        // SELECT S.S_NO, S.NAME, S.ADDRESS, C.CLASS_NAME, C.TEACHER
        // FROM STUDENT S,
        //      STUDENT_CLASS SC,
        //      CLASS C
        // WHERE S.S_NO = SC.S_NO
        //   AND SC.C_NO = C.C_NO
        //   AND S.NAME = '홍길동';

        val targetStudentName = "홍길동"

        students
            .filter { it.name == targetStudentName } // WHERE S.NAME = '홍길동'
            .flatMap { student -> // FROM STUDENT S
                studentClasses
                    .filter { sc -> sc.sNo == student.sNo } // JOIN STUDENT_CLASS SC ON S.S_NO = SC.S_NO
                    .flatMap { sc -> // 이제 student와 sc (수강정보)를 알고 있음
                        classes
                            .filter { cls -> cls.cNo == sc.cNo } // JOIN CLASS C ON SC.C_NO = C.C_NO
                            .map { cls -> // SELECT S.S_NO, S.NAME, S.ADDRESS, C.CLASS_NAME, C.TEACHER
                                EnrollmentInfo(
                                    sNo = student.sNo,
                                    studentName = student.name,
                                    studentAddress = student.address,
                                    className = cls.className,
                                    teacherName = cls.teacher
                                )
                            }
                    }
            }
            .forEach { println(it) }
    }

    @Test
    fun testAssociatedMap() {

        val targetStudentName = "홍길동"

        // 수업 정보를 cNo를 키로 하는 Map으로 변환 (조회 성능 향상)
        val classMap = classes.associateBy { it.cNo }

        // 학생-수업 관계 정보를 sNo를 키로 하고, cNo 리스트를 값으로 하는 Map으로 변환 (한 학생이 여러 수업을 들을 수 있으므로)
        // 여기서는 studentClasses가 sNo당 cNo가 하나씩만 있다고 가정하고 단순화 (만약 여러 개면 groupBy 사용)
        val studentClassMap = studentClasses.associate { it.sNo to it.cNo } // 여기서는 1:1 매핑 가정
        // 만약 한 학생이 여러 수업을 듣는다면:
        // val studentToClassNosMap = studentClasses.groupBy({ it.sNo }, { it.cNo })

        students
            .filter { it.name == targetStudentName }
            .flatMap { student ->
                // student.sNo로 studentClassMap에서 cNo를 찾음
                val cNo = studentClassMap[student.sNo]
                if (cNo != null) {
                    // cNo로 classMap에서 ClassInfo를 찾음
                    val classInfo = classMap[cNo]
                    if (classInfo != null) {
                        listOf( // flatMap은 각 요소에 대해 컬렉션(여기서는 List)을 반환해야 함
                            EnrollmentInfo(
                                sNo = student.sNo,
                                studentName = student.name,
                                studentAddress = student.address,
                                className = classInfo.className,
                                teacherName = classInfo.teacher
                            )
                        )
                    } else {
                        emptyList() // 매칭되는 ClassInfo가 없으면 빈 리스트 반환
                    }
                } else {
                    emptyList() // 매칭되는 StudentClass 정보가 없으면 빈 리스트 반환
                }
            }
            .forEach { println(it) }
    }
}